src/Microsoft.ML.LightGbm/LightGbmArguments.cs

```diff
@@ -1,678 +1,225 @@
-                [TlcModule.SweepableDiscreteParam("RegAlpha", new object[] { 0f, 0.5f, 1f })]
-                public double L1Regularization = 0;
-
-                /// <summary>
-                /// Controls the balance of positive and negative weights in <see cref="LightGbmBinaryClassificationTrainer"/>.
-                /// </summary>
-                /// <value>
-                /// This is useful for training on unbalanced data. A typical value to consider is sum(negative cases) / sum(positive cases).
-                /// </value>
-                [Argument(ArgumentType.AtMostOnce,
-                    HelpText = "Control the balance of positive and negative weights, useful for unbalanced classes." +
-                    " A typical value to consider: sum(negative cases) / sum(positive cases).",
-                    ShortName = "ScalePosWeight")]
-                public double WeightOfPositiveExamples = 1;
-
-                internal virtual IBoosterParameter CreateComponent(IHostEnvironment env) => new TreeBooster(this);
-
-                IBoosterParameter IComponentFactory<IBoosterParameter>.CreateComponent(IHostEnvironment env) => CreateComponent(env);
-            }
+                var attribute = field.GetCustomAttribute<ArgumentAttribute>(false);
 
-            internal TreeBooster(Options options)
-                : base(options)
-            {
-                Contracts.CheckUserArg(BoosterParameterOptions.MinimumSplitGain >= 0, nameof(BoosterParameterOptions.MinimumSplitGain), "must be >= 0.");
-                Contracts.CheckUserArg(BoosterParameterOptions.MinimumChildWeight >= 0, nameof(BoosterParameterOptions.MinimumChildWeight), "must be >= 0.");
-                Contracts.CheckUserArg(BoosterParameterOptions.SubsampleFraction > 0 && BoosterParameterOptions.SubsampleFraction <= 1, nameof(BoosterParameterOptions.SubsampleFraction), "must be in (0,1].");
-                Contracts.CheckUserArg(BoosterParameterOptions.FeatureFraction > 0 && BoosterParameterOptions.FeatureFraction <= 1, nameof(BoosterParameterOptions.FeatureFraction), "must be in (0,1].");
-                Contracts.CheckUserArg(BoosterParameterOptions.L2Regularization >= 0, nameof(BoosterParameterOptions.L2Regularization), "must be >= 0.");
-                Contracts.CheckUserArg(BoosterParameterOptions.L1Regularization >= 0, nameof(BoosterParameterOptions.L1Regularization), "must be >= 0.");
-                Contracts.CheckUserArg(BoosterParameterOptions.WeightOfPositiveExamples > 0, nameof(BoosterParameterOptions.WeightOfPositiveExamples), "must be >= 0.");
-            }
+                if (attribute == null)
+                    continue;
 
-            internal override void UpdateParameters(Dictionary<string, object> res)
-            {
-                base.UpdateParameters(res);
-                res["boosting_type"] = Name;
+                res[LightGbmInterfaceUtils.GetOptionName(field.Name)] = field.GetValue(BoosterOptions);
             }
```

Comment:
Do we need to check if key `LightGbmInterfaceUtils.GetOptionName(field.Name)` exists?

Reply:
So LightGbmInterfaceUtils.GetOptionName does not have a dictionary that it works from -- it just converts the field name to lower case plus some underscoring. So there should be no need to check for a key. There could already be a key in the res dictionary - but thats ok as we are overriding it and like you said, this is an Update function.


---
In reply to: [265815549](https://github.com/dotnet/machinelearning/pull/2948#discussion_r265815549) [](ancestors = 265815549)

Reply:
Ill mark this as pending as I want to make sure I answered your question.

---
In reply to: [266178686](https://github.com/dotnet/machinelearning/pull/2948#discussion_r266178686) [](ancestors = 266178686,265815549)